### PR TITLE
please merge this PR

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by iejMac** [comment]
openai this moron submitted this huge PR, can you summarize it for me?

**Comment by iejMac** [comment]
openai can you try it again?

**Comment by rom1504** [comment]
openai hi

**Comment by iejMac** [comment]
openai can you give feedback?

**Comment by jkcoxson** [comment]
Wow this is really cool!

**Comment by iejMac** [comment]
openai try again

**Comment by iejMac** [comment]
openai check this out

**Comment by iejMac** [comment]
openai check it out

**Comment by iejMac** [comment]
openai last try
